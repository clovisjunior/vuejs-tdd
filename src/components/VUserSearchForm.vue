<script>
export default {
    name: 'UserSearchForm',
    data() {
        return {
            username: ''
        }
    }
}
</script>

<template>
    <form @submit.prevent="$emit('submitted', username)">
        <el-input
            placeholder="Pesquise o usuÃ¡rio"
            v-model="username"
            class="search-form__input"
        >
            <el-button
                class="search-form__button"
                native-type="submit"
                slot="append"
                icon="el-icon-search"
            ></el-button>
        </el-input>        
    </form>
</template>
