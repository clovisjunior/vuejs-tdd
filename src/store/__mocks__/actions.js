import userFixture from '../../../tests/unit/fixtures/user'

export default {
    SEARCH_USER: jest.fn().mockResolvedValue(userFixture)
}