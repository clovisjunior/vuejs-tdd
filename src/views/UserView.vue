<script>
import { mapState } from 'vuex'
import VUserSearchForm from '@/components/VUserSearchForm'
import VUserProfile from '@/components/VUserProfile'

export default {
    name: 'UserView',
    components: {
        VUserSearchForm,
        VUserProfile
    },
    methods: {
        searchUser(username) {
            this.$store.dispatch('SEARCH_USER', { username })
        }
    },
    computed: {
        ...mapState({
            user: 'user'
        })
    }
}
</script>


<template>
    <div>
        <VUserSearchForm
            @submitted="searchUser"/>
        <VUserProfile :user="user"/>
    </div>
</template>


