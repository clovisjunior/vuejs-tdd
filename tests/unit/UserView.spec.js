jest.mock('@/store/actions')
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import UserView from '@/views/UserView'
import VUserSearchForm from '@/components/VUserSearchForm'
import VUserProfile from '@/components/VUserProfile'
import initialState from '@/store/state'
import actions from '@/store/actions'
import userFixture from './fixtures/user'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('UserView', () => {
    let state

    const build = () => {
        const wrapper = shallowMount(UserView, {
            localVue,
            store: new Vuex.Store({
                state,
                actions
            })
        })

        return {
            wrapper,
            userSearchForm: () => wrapper.find(VUserSearchForm),
            userProfile: () => wrapper.find(VUserProfile)
        }
    }

    beforeEach( () => {
        jest.resetAllMocks()
        state = { ...initialState }
    })

    it('renders the component', () => {
        const { wrapper } = build()
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders main child components', () => {
        const { userSearchForm, userProfile } = build()

        expect(userSearchForm().exists()).toBe(true)
        expect(userProfile().exists()).toBe(true)

    })

    it('passes a binded user prop to user profile component', () => {
        state.user = userFixture
        const { userProfile } = build()

        expect(userProfile().vm.user).toBe(state.user)
    })

    it('searches for a user when received "submitted"', () => {

        const expecteduser = 'clovisjunior'
        const { userSearchForm } = build()

        userSearchForm().vm.$emit('submitted', expecteduser)

        expect(actions.SEARCH_USER).toHaveBeenCalled()
        expect(actions.SEARCH_USER.mock.calls[0][1]).toEqual({
            username: expecteduser
        })
    })
})