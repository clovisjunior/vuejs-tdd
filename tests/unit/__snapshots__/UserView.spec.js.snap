// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserView renders the component 1`] = `
<div>
  <vusersearchform-stub></vusersearchform-stub>
  <vuserprofile-stub user="[object Object]"></vuserprofile-stub>
</div>
`;
